### PR TITLE
COMP: Address ContourExtractor2DImageFilter valgrind defects

### PR DESCRIPTION
**Pull request by Leengit**
Address Issue #2426.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx [hunk 1]
```diff
@@ -486,7 +486,8 @@ ContourExtractor2DImageFilter<TInputImage>::AddSegment(VertexType from, VertexTy
         // Now remove the old end of 'head' from the ends map and add
         // the new end.
         contourData.m_ContourEnds.erase(newHead);
-        contourData.m_ContourEnds.insert(VertexToContourContainerIteratorMapKeyValuePair(head->back(), head));
+        const VertexToContourContainerIteratorMapKeyValuePair newEnd(head->back(), head);
+        contourData.m_ContourEnds.insert(newEnd);
```
Comment (blowekamp) [inline Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx:490]: Hello, Thank you for addressing the issue about the valgrind defects I created. 

I'm sure you PR is just fine the way it is... but I thought I'd give one of those extra suggestions :)
Consider:

`contourData.m_ContourEnds.emplace(head->back(), head)`

**Review by dzenanz — APPROVED** [review]
Looks good at a glance. Brad's suggestion sounds helpful.

File: Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx (hunk 1 above)
Comment (Leengit) [reply Modules/Filtering/Path/include/itkContourExtractor2DImageFilter.hxx]: Good, thank you.  I have now done that.

**Comment by Leengit** [comment]
In addition to using `emplace`, with this latest push I have also changed a `std::unordered_map`  to a `std::map`.  Although the former, is _O(n)_ and the latter is _O(n lg n)_, only the latter guarantees that an iterator set before an erase or insert will remain valid afterwards, and we need that.  Maybe this is what valgrind was trying to tell us?

**Comment by jhlegarreta** [comment]
@Leengit thanks for giving this a try. Looks like the class is giving a lot of work :pensive: .

I've been noticing that the test is also failing for some of the nightly builds, e.g.:
https://open.cdash.org/test/367698202

Just in case the information might be relevant, I had a look at the failure report: it looks like the  `AddSegment` method is triggering a segfault. The proposed changes to the implementation file in this PR happen in that method. You may want to have a look at the error report in the link aboce if you have not done so/if you get a chance. Maybe the information provided helps to better understand the root of the problem.

**Comment by Leengit** [comment]
@jhlegarreta -- I am hoping that the switch from `std::unordered_map` to `std::map` will address the segmentation faults.  But if you notice otherwise, please submit a follow up issue.

**Comment by jhlegarreta** [comment]
@Leengit Valgrind is reporting no defects now !! :fireworks: Thanks for having fixed this.

**Pull request merged**